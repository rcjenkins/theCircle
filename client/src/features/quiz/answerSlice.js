import { createSlice } from '@reduxjs/toolkit';

export const answersSlice = createSlice({
  name: 'answers',
  initialState: {
    answers: null,
    isFetching: false,
    error: null,
  },
  reducers: {
    requestSaveAnswers: (state) => {
      state.response = null;
      state.isFetching = true;
      state.error = null;
    },
    successSaveAnswers: (state, action) => {
      state.response = action.payload;
      state.isFetching = false;
      state.error = null;
    },
    failedSaveAnswers: (state, action) => {
      state.response = null;
      state.isFetching = false;
      state.error = action.payload.message;
    },
  },
});

// saving the answers

export const saveAnswers = () => (dispatch, getState) => {
  const {
    requestSaveAnswers,
    successSaveAnswers,
    failedSaveAnswers,
  } = answersSlice.actions;
  const { quiz: quizSlice } = getState();
  const { quiz, answers } = quizSlice;

  // wrapping the simple answers array for sending to API
  const payload = {
    quizData: {
      quizId: quiz.id.toString(),
      questionAnswers: answers,
    },
  };
  const apiEndpoint = window.awsAPI.apiEndpoint + '?quizId=' + quiz.id;
  dispatch(requestSaveAnswers());
  fetch(apiEndpoint, {
    method: 'POST',
    cache: 'no-cache',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  })
    .then((response) => response.json())
    .then((json) => dispatch(successSaveAnswers(json)))
    .catch((error) => dispatch(failedSaveAnswers(error)));
};

export const selectAnswers = (state) => state.answers;

export default answersSlice.reducer;
