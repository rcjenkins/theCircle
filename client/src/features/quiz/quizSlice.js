import { createSlice } from '@reduxjs/toolkit';

export const quizSlice = createSlice({
  name: 'quiz',
  initialState: {
    answers: [],
    quiz: null,
    isFetching: false,
    error: null,
    currentQuestionId: null,
  },
  reducers: {
    requestLoadQuiz: (state) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.answers = [];
      state.quiz = null;
      state.isFetching = true;
      state.error = null;
      state.currentQuestionId = null;
    },
    successLoadQuiz: (state, action) => {
      state.answers = [];
      state.quiz = action.payload;
      state.isFetching = false;
      state.error = null;
      // make the currentQuestionId the first question id after load as some quizzes may not start at 0
      state.currentQuestionId = state.quiz.questions[0].id;
    },
    failedLoadQuiz: (state, action) => {
      state.answers = [];
      state.quiz = null;
      state.isFetching = false;
      state.error = action.payload.message;
      state.currentQuestionId = null;
    },
    addAnswer: (state, action) => {
      state.answers.push({
        id: state.currentQuestionId.toString(),
        answers: action.payload.answers,
      });
      state.currentQuestionId = action.payload.nextQuestionId;
    },
    removeAnswer(state) {
      const lastAnswer = state.answers.pop();
      if (lastAnswer) {
        state.currentQuestionId = lastAnswer.id;
      }
    },
  },
});

// I like to hide away the actual actions this stops potential typos in the parameter object

export const addAnswer = (answers, nextQuestionId) => (dispatch) => {
  const { addAnswer } = quizSlice.actions;
  dispatch(addAnswer({ answers, nextQuestionId }));
};

export const removeAnswer = () => (dispatch) => {
  const { removeAnswer } = quizSlice.actions;
  dispatch(removeAnswer());
};

// loading quiz data using API

export const loadQuiz = () => (dispatch) => {
  const {
    requestLoadQuiz,
    successLoadQuiz,
    failedLoadQuiz,
  } = quizSlice.actions;
  dispatch(requestLoadQuiz());

  //const localEndpoint = `${window.location.origin}/quiz-0.json`;
  const apiEndpoint = window.awsAPI.apiEndpoint + '?quizId=0';
  fetch(apiEndpoint)
    .then((response) => response.json())
    .then((json) => dispatch(successLoadQuiz(json)))
    .catch((error) => dispatch(failedLoadQuiz(error)));
};

export const selectQuiz = (state) => state.quiz;

export default quizSlice.reducer;
