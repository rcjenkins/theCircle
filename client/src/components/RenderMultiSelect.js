import { Input, Alert, Label, FormGroup, Button } from 'reactstrap';
import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { selectQuiz, addAnswer } from '../features/quiz/quizSlice';

const RenderMultiSelect = () => {
  const [state, setState] = useState({ selected: [] });

  const dispatch = useDispatch();
  const quizData = useSelector(selectQuiz);
  const question = quizData.quiz.questions.find(
    (element) => element.id === quizData.currentQuestionId,
  );

  const renderOption = (answer) => {
    return (
      <option key={answer.id} id={answer.id}>
        {answer.text}
      </option>
    );
  };

  const renderAlert = () => {
    return (
      <Alert color="danger">
        Please select one or more answers from the select.
      </Alert>
    );
  };

  const handleChange = (event) => {
    let options = event.target.options;
    let value = [];
    for (var i = 0, l = options.length; i < l; i++) {
      if (options[i].selected) {
        value.push(options[i].id);
      }
    }

    setState({ selected: value });
  };

  const handleClick = () => {
    if (state.selected.length > 0) {
      dispatch(addAnswer(state.selected, question.nextQuestionId));
    } else {
      setState({ ...state, viewAlert: true });
    }
  };

  return (
    <>
      <FormGroup>
        <Label id="exampleSelectLabel" htmlFor="exampleSelect">
          {question.title}
        </Label>
        <Input
          type="select"
          name="select"
          id="exampleSelect"
          size={`${question.answers.length}`}
          multiple
          onChange={handleChange}
        >
          {question.answers.map((answer) => {
            return renderOption(answer);
          })}
        </Input>
      </FormGroup>
      <FormGroup>
        <Button color="primary" onClick={() => handleClick()}>
          Continue
        </Button>
      </FormGroup>

      {state.viewAlert && renderAlert()}
    </>
  );
};

export default RenderMultiSelect;
