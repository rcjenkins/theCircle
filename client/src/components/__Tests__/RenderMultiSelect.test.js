import React from 'react';
import RenderMultiSelect from '../RenderMultiSelect';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { selectQuiz } from '../../features/quiz/quizSlice';

jest.mock('../../features/quiz/quizSlice');

it('RenderMultiSelect renders correctly', () => {
  selectQuiz.mockReturnValue({
    answers: [],
    isFetching: false,
    error: null,
    currentQuestionId: 1,
    quiz: {
      id: 0,
      title: 'Daily Engagement Service',
      questions: [
        {
          id: 1,
          title: 'Great what contributed to your success today?',
          type: 'multiSelect',
          nextQuestionId: 4,
          answers: [
            {
              id: 2,
              text: 'My colleagues.',
            },
            {
              id: 3,
              text: 'I felt empowered to get my job done.',
            },
            {
              id: 4,
              text: 'I had access to all of the tools I needed to smash it.',
            },
            {
              id: 5,
              text: 'I could focus on my most important tasks.',
            },
            {
              id: 6,
              text: 'I knew exactly what I was doing so just nailed it.',
            },
            {
              id: 7,
              text: "I don't know.  I was just on a role.",
            },
          ],
        },
      ],
    },
  });

  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);

  const initialState = {};
  const store = mockStore(initialState);
  const tree = renderer
    .create(
      <Provider store={store}>
        <RenderMultiSelect />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
