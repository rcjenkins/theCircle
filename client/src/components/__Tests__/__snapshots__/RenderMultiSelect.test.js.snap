// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderMultiSelect renders correctly 1`] = `
Array [
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="exampleSelect"
      id="exampleSelectLabel"
    >
      Great what contributed to your success today?
    </label>
    <select
      className="form-control"
      id="exampleSelect"
      multiple={true}
      name="select"
      onChange={[Function]}
      size="6"
    >
      <option
        id={2}
      >
        My colleagues.
      </option>
      <option
        id={3}
      >
        I felt empowered to get my job done.
      </option>
      <option
        id={4}
      >
        I had access to all of the tools I needed to smash it.
      </option>
      <option
        id={5}
      >
        I could focus on my most important tasks.
      </option>
      <option
        id={6}
      >
        I knew exactly what I was doing so just nailed it.
      </option>
      <option
        id={7}
      >
        I don't know.  I was just on a role.
      </option>
    </select>
  </div>,
  <div
    className="form-group"
  >
    <button
      aria-label={null}
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      Continue
    </button>
  </div>,
]
`;
